### PR TITLE
Move apiserver auth sa and user under component helpers

### PR DESCRIPTION
**Pull request by ingvagabund**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Move definitions of `k8s.io/component-helpers/auth/serviceaccount` and `k8s.io/apiserver/pkg/authentication/user` package from  `k8s.io/apiserver` under `k8s.io/component-helpers/auth`. Both dependencies are used in at least two components (kube-apiserver and kubectl). Moving them out of `k8s.io/apiserver` will allow `k8s.io/kubernetes/pkg/registry/rbac/reconciliation` `k8s.io/kubernetes/pkg/registry/rbac/validation` to be moved under `k8s.io/component-helpers/auth` as well which will untangle dependency of `kubectl` on `k8s.io/kubernetes/pkg/registry/rbac/reconciliation`. This PR is a subset of minimal changes contained in https://github.com/kubernetes/kubernetes/pull/95761 to easy the migration of the helper code.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/edit#bookmark=id.mbp70qxbod54

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

**TODO**:
- [x] - create new root `auth` directory
- [x] - move `k8s.io/apiserver/pkg/authentication/serviceaccount` under `k8s.io/component-helpers/auth`
- [x] - move `k8s.io/apiserver/pkg/authentication/user` under `k8s.io/component-helpers/auth`

### CODE REVIEW

File: cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ import (
 	"k8s.io/api/core/v1"
 	rbac "k8s.io/api/rbac/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/component-helpers/auth/user"
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go:26]: /approve
for kubeadm, thanks!

**Comment by ingvagabund** [comment]
Just running `gofmt` and updating `staging/publishing/rules.yaml` to make `pull-kubernetes-verify` happy.

**Comment by ingvagabund** [comment]
@liggitt @deads2k @tallclair @yguo0905 @dims PTAL

**Comment by ingvagabund** [comment]
@liggitt @deads2k @tallclair @yguo0905 @dims PTAL for approval

**Comment by deads2k** [comment]
moving user.Info into component-helpers seems unusual and unnecessary.  

I'm not completely clear on what kubectl uses it for, but rather than move this way, I would suggest letting kubectl depend on k8s.io/apiserver if necessary until you can unwind that aspect of it.  I think that compatible Interfaces can in any package if you need a copy.

**Comment by liggitt** [comment]
Looking at this again, I don't think kubectl reconcile actually needs user.Info interfaces. My first attempt just moved entire RBAC packages, and `pkg/registry/rbac/validation` had the AuthorizationRuleResolver interface, which required the user.Info interface. If we split out the policy comparison bits of that package, I think we can be way more judicious:

I think the following things would make sense to move to component-helpers:

pkg/registry/rbac/validation:
- policy_compact_test.go
- policy_compact.go
- policy_comparator_test.go
- policy_comparator.go

pkg/apis/rbac/v1:
- SortableRuleSlice

pkg/registry/rbac/reconciliation

**Comment by soltysh** [comment]
> I don't think kubectl reconcile actually needs user.Info interfaces. 

It does not from what I just checked. I'm definitely in favor of minimal necessary changes that Jordan pointed out in previous comment.

**Comment by ingvagabund** [comment]
> I think the following things would make sense to move to component-helpers:
>
> pkg/registry/rbac/validation:
>
> policy_compact_test.go
> policy_compact.go
> policy_comparator_test.go
> policy_comparator.go
>
> pkg/apis/rbac/v1:
> 
> SortableRuleSlice
>
> pkg/registry/rbac/reconciliation

I checked for minimal dependencies to drop `k8s.io/kubernetes/pkg/registry/rbac/reconciliation` dependency from `pkg/kubectl`. Only `policy_comparator.go` and `policy_comparator_test.go` from `pkg/registry/rbac/validation` are needed to bring `Covers` helper. Neither is `SortableRuleSlice` as it's imported in `policy_comparator_test.go`.

I updated the original PR from which this PR came. Moving only `pkg/registry/rbac/reconciliation` and `[policy_comparator.go, policy_comparator_test.go]` from `pkg/registry/rbac/validation`.

This PR can be closed and the discussion moved to https://github.com/kubernetes/kubernetes/pull/95761.

/close

**Pull request closed without merging**